src/FAudio_gstreamer.c

```diff
@@ -0,0 +1,850 @@
+			{
+				if (!FAudio_GSTREAMER_DecodeBlock(voice, buffer, curBlock, maxBytes))
+				{
+					error = 1;
+					goto done;
+				}
+			}
+			break;
+		}
+
+		byteOffset -= gstreamer->blockSizes[curBlock];
+		maxBytes -= gstreamer->blockSizes[curBlock];
+		if(maxBytes == 0)
+			break;
+	}
+
+	if (curBlock >= voice->src.bufferList->bufferWMA.PacketCount || maxBytes == 0)
+	{
+		goto done;
+	}
+
+	/* If we're in a different block from last time, decode! */
+	if (curBlock == gstreamer->curBlock)
+	{
+		convertCache = gstreamer->convertCache;
+	}
+	else if (curBlock == gstreamer->prevBlock)
+	{
+		convertCache = gstreamer->prevConvertCache;
+	}
+	else
+	{
+		FAudio_assert(0 && "Somehow got an undecoded curBlock!");
+	}
+
+	/* Copy to decodeCache, finally. */
+	siz = FAudio_min(SIZE_FROM_DST(samples), gstreamer->blockSizes[curBlock] - byteOffset);
+	if (convertCache)
+	{
+		FAudio_memcpy(
+			decodeCache,
+			convertCache + byteOffset,
+			siz
+		);
+	}
+	else
+	{
+		FAudio_memset(
+			decodeCache,
+			0,
+			siz
+		);
+	}
+
+	/* Increment pointer, decrement remaining sample count */
+	decodeCache += siz / sizeof(float);
+	samples -= SAMPLES_FROM_DST(siz);
+	curBufferOffset += SAMPLES_FROM_DST(siz);
+
+done:
+	if (FAudio_GSTREAMER_CheckForBusErrors(voice))
+	{
+		LOG_ERROR(
+			voice->audio,
+			"%s",
+			"Got a bus error after decoding!"
+		)
+
+		error = 1;
+	}
+
+	/* If the cache isn't filled yet, keep decoding! */
+	if (samples > 0)
+	{
+		if (	!error &&
+			curBlock < voice->src.bufferList->bufferWMA.PacketCount - 1	)
+		{
+			goto decode;
+		}
+
+		/* out of stuff to decode, write blank and exit */
+		FAudio_memset(decodeCache, 0, SIZE_FROM_DST(samples));
+	}
+
+	LOG_FUNC_EXIT(voice->audio)
+}
+
+void FAudio_GSTREAMER_end_buffer(FAudioSourceVoice *pSourceVoice)
+{
+	FAudioGSTREAMER *gstreamer = pSourceVoice->src.gstreamer;
+
+	LOG_FUNC_ENTER(pSourceVoice->audio)
+
+	pSourceVoice->audio->pFree(gstreamer->blockSizes);
+	gstreamer->blockSizes = NULL;
+
+	LOG_FUNC_EXIT(pSourceVoice->audio)
+}
+
+static void FAudio_GSTREAMER_NewDecodebinPad(GstElement *decodebin,
+		GstPad *pad, gpointer user)
+{
+	FAudioGSTREAMER *gstreamer = user;
+	GstPad *ac_sink;
+
+	ac_sink = gst_element_get_static_pad(gstreamer->resampler, "sink");
+	if (GST_PAD_IS_LINKED(ac_sink))
+	{
+		gst_object_unref(ac_sink);
+		return;
+	}
+
+	gst_pad_link(pad, ac_sink);
+
+	gst_object_unref(ac_sink);
+}
+
+static uint8_t gstreamerWasInit = 0;
+
+uint32_t FAudio_GSTREAMER_init(FAudioSourceVoice *pSourceVoice, uint32_t type)
+{
+	FAudioGSTREAMER *result;
+	GstElement *decoder = NULL, *converter = NULL;
+	GstCaps *caps;
+	int version;
+	GstBuffer *codec_data;
+	size_t codec_data_size;
+	uint8_t *extradata;
+	uint8_t fakeextradata[16];
+	GValue gval = G_VALUE_INIT;
+	GstPad *decoder_sink;
+	GstEvent *event;
+
+	LOG_FUNC_ENTER(pSourceVoice->audio)
+
+	/* Init GStreamer statically. The docs tell us not to exit, so I guess
+	 * we're supposed to just leak!
```

Comment:
If you want you can call `gst_deinit()` at the very end, but you can't call `gst_init()` or any other GStreamer API again afterwards